tasks/DroneFlightControlTask.cpp

```diff
@@ -0,0 +1,440 @@
+    for(int i=0;i<16;i++)
+    {
+        wp.commandList[i] = cmd_waypoint.commandList[i];
+        wp.commandParameter[i] = cmd_waypoint.commandParameter[i];
+    }
+    return wp;
+}
+
+power_base::BatteryStatus DroneFlightControlTask::getBatteryStatus() const
+{
+    auto djiBattery = mVehicle->broadcast->getBatteryInfo();
+
+    power_base::BatteryStatus battery;
+    battery.time = base::Time::fromMilliseconds(mVehicle->broadcast->getTimeStamp().time_ms);
+    battery.current = djiBattery.current;
+    battery.voltage = djiBattery.voltage;
+    battery.charge = static_cast<float>(djiBattery.percentage) / 100;
+    return battery;
+}
+
+base::samples::RigidBodyState DroneFlightControlTask::getRigidBodyState() const
+{
+    base::samples::RigidBodyState cmd;
+    cmd.time = base::Time::fromMilliseconds(mVehicle->broadcast->getTimeStamp().time_ms);
+    cmd.orientation.w() = mVehicle->broadcast->getQuaternion().q0;
+    cmd.orientation.x() = mVehicle->broadcast->getQuaternion().q1;
+    cmd.orientation.y() = mVehicle->broadcast->getQuaternion().q2;
+    cmd.orientation.z() = mVehicle->broadcast->getQuaternion().q3;
+    cmd.velocity.x() = mVehicle->broadcast->getVelocity().x;
+    cmd.velocity.y() = mVehicle->broadcast->getVelocity().y;
+    cmd.velocity.z() = mVehicle->broadcast->getVelocity().z;
+    cmd.angular_velocity.x() = mVehicle->broadcast->getAngularRate().x;
+    cmd.angular_velocity.y() = mVehicle->broadcast->getAngularRate().y;
+    cmd.angular_velocity.z() = mVehicle->broadcast->getAngularRate().z;
+    // Get GPS position information
+    Telemetry::GPSInfo gpsInfo = mVehicle->broadcast->getGPSInfo();
+    // Convert position data from GPS to NWU
+    gps_base::Solution solution;
+    solution.time = cmd.time;
+    solution.latitude = gpsInfo.latitude;
+    solution.longitude = gpsInfo.longitude;
+    solution.altitude = gpsInfo.HFSL;
+    // Get position
+    gps_base::UTMConverter gpsSolution;
+    base::samples::RigidBodyState gpsPosition = gpsSolution.convertToNWU(solution);
+    cmd.position = gpsPosition.position;
+
+    return cmd;
+}
+
+bool DroneFlightControlTask::setUpSubscription(int pkgIndex, int freq,
+                                               Telemetry::TopicName topicList[],
+                                               uint8_t topicSize)
+{
+    if (mVehicle)
+    {
+        /*! Telemetry: Verify the subscription*/
+        ACK::ErrorCode subscribeStatus;
+        subscribeStatus = mVehicle->subscribe->verify(mFunctionTimeout);
+        if (ACK::getError(subscribeStatus) != ACK::SUCCESS)
+        {
+            ACK::getErrorCodeMessage(subscribeStatus, __func__);
+            return false;
+        }
+
+        bool enableTimestamp = false;
+        bool pkgStatus = mVehicle->subscribe->initPackageFromTopicList(
+            pkgIndex, topicSize, topicList, enableTimestamp, freq);
+        if (!(pkgStatus))
+            return pkgStatus;
+
+        usleep(5000);
```

Comment:
Why the 5ms sleep ?

Reply:
i don't know, I just followed the last version of the sample code. Now, that we are in the v3.9 of the sdk, I will remove that.